tools/fft.c

```diff
@@ -0,0 +1,381 @@
+			window->in_buffer, window->out_buffer, FFTW_MEASURE);
+	if (window->plan == NULL) {
+		goto cleanup_out;
+	}
+	return window;
+
+cleanup_out:
+		fftw_free(window->out_buffer);
+cleanup_in:
+		free(window->in_buffer);
+cleanup_window:
+		free(window);
+		return NULL;
+}
+
+void destroy_sample_window(struct sample_window *window)
+{
+	fftw_destroy_plan(window->plan);
+	fftw_free(window->out_buffer);
+	free(window->in_buffer);
+	free(window);
+}
+
+static inline bool window_is_full(struct sample_window *window)
+{
+	return window->sample_count == window->buffer_capacity;
+}
+
+unsigned add_sample_to_window(struct sample_window *window, double sample)
+{
+	if (window->sample_count < window->buffer_capacity) {
+		window->in_buffer[window->sample_count] = sample;
+		window->sample_count++;
+	}
+	if (window_is_full(window)) {
+		if (!window->transformed) {
+			fftw_execute(window->plan);
+		}
+		return 1;
+	}
+	return 0;
+}
+
+static inline void mod_welch_average(struct channel_data *channel,
+		unsigned window_value, struct sample_window *window, int mod)
+{
+	for (uint32_t i = 0; i < window->out_capacity; i++) {
+		double real, imag, mag, addition;
+		real = window->out_buffer[i][0];
+		imag = window->out_buffer[i][1];
+		mag = real * real + imag * imag;
+		addition = mag / window_value / channel->welch_window_count;
+
+		channel->welch_output[i] += mod * addition;
+	}
+}
+
+static inline void add_to_welch_average(struct channel_data *channel,
+		unsigned window_value, struct sample_window *window)
+{
+	mod_welch_average(channel, window_value, window, 1);
+}
+
+static inline void subtract_from_welch_average(struct channel_data *channel,
+		unsigned window_value, struct sample_window *window)
+{
+	mod_welch_average(channel, window_value, window, -1);
+}
+
+void welch_method(struct channel_data *channel)
+{
+	// Unless we use a non-square window function, the window value is easy
+	// Otherwise, sum the windowed value squared over the entire window
+	unsigned window_value = channel->window_length;
+	unsigned out_length = channel->window_length / 2 + 1;
+	struct sample_window *window;
+	if (channel->welch_output == NULL) {
+		// Populate from every full window if this is the first run
+		channel->welch_output = calloc(out_length,
+				sizeof(*channel->welch_output));
+
+		for (unsigned i = 0; pfifo_peek_back(channel->windows, i,
+					(void**) &window); i++) {
+			if (window_is_full(window)) {
+				add_to_welch_average(channel, window_value, window);
+			}
+		}
+	} else {
+		// Get the most recent full window
+		for (unsigned i = 0; pfifo_peek_back(channel->windows, i,
+					(void**) &window); i++) {
+			if (window_is_full(window)) {
+				break;
+			}
+		}
+		add_to_welch_average(channel, window_value, window);
+	}
+}
+
+void print_welch_output(struct channel_data *channel)
+{
+	unsigned out_length = channel->window_length / 2 + 1;
+	static uint64_t output_index = 0; // to differentiate output FFTs
```

Comment:
Would it be worth storing the output index in the `struct channel_data`, so that each channel would have a separate index?

Reply:
This seems like a good idea.